src/main/java/com/commercetools/sync/commons/helpers/GenericCustomActionBuilder.java

```diff
@@ -2,57 +2,76 @@
-     * the given resource {@code T} (which currently could either be a {@link Category} or a {@link Channel}).
+     * the given resource {@code T}. If the resource that has the custom fields is a secondary resource (e.g. price or
+     * asset) and not a primary resource (e.g Category, Product, Channel, etc..), the {@code variantId} and the
+     * {@code objectId} will be used to identify this secondary resource within its container.
      *
+     * @param variantId an optional field which could be used to identify the variant that holds the a resource
+     *                  (e.g. asset) which has the custom fields.
+     * @param objectId  an optional field which could be used to identify the id of the resource
+     *                  (e.g. asset, price, etc..) which has the custom fields.
      * @return a setCustomType update action that removes the custom type from the resource it's requested on.
      */
     @Nonnull
-    public abstract UpdateAction<T> buildRemoveCustomTypeAction();
+    public abstract UpdateAction<T> buildRemoveCustomTypeAction(@Nullable final Integer variantId,
+                                                                @Nullable final String objectId);
 
     /**
-     * Creates a CTP "setCustomType" update action on the given resource {@code T} (which currently could either
-     * be a {@link Category} or a {@link Channel}).
+     * Creates a CTP "setCustomType" update action on the given resource {@code T}. If the resource that has the custom
+     * fields is a secondary resource (e.g. Price or asset) and not a primary resource (e.g Category, Product, Channel,
+     * etc..), the {@code variantId} and the {@code objectId} will be used to identify this secondary resource within
+     * its container.
      *
-     * @param customTypeId       the id of the new custom type.
+     * @param variantId           an optional field which could be used to identify the variant that holds the a
+     *                            resource (e.g. asset) which has the custom fields.
+     * @param objectId            an optional field which could be used to identify the id of the resource
+     *                            (e.g. asset, price, etc..) which has the custom fields.
+     * @param customTypeId        the id of the new custom type.
      * @param customFieldsJsonMap the custom fields map of JSON values.
      * @return a setCustomType update action of the type of the resource it's requested on.
      */
     @Nonnull
-    public abstract UpdateAction<T> buildSetCustomTypeAction(@Nullable final String customTypeId,
+    public abstract UpdateAction<T> buildSetCustomTypeAction(@Nullable final Integer variantId,
+                                                             @Nullable final String objectId,
+                                                             @Nullable final String customTypeId,
```

Comment:
> Actually, i'm not sure nullable has any sense in this context

It definitely has since it could be passed as `null` from the user, since it's just passed in a user-built draft. 

I was also thinking whether it makes sense to do a null check before calling this method. But then I saw that the only place where the method is called, there is a try-catch. And this should catch the NPE. I just noticed noticed right now though (thanks!) that there is no NPE in the catch statement. Need to figure out how to handle this.

Reply:
ok

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/253/files#diff-e4a21b83897a2eafe26a012b94742081R55

and unit tests: https://github.com/commercetools/commercetools-sync-java/pull/253/commits/9685eae4609c7e0a0f8de63fc5bbc3dea06c8235, https://github.com/commercetools/commercetools-sync-java/pull/253/commits/1b7097181ac95ccbc21be87884298df4e2656399